### ISSUE TITLE
`[ERROR] Metadata copy process failed with code: 1` is thrown when using `native:metadata-copy` in a multi-module Maven project

### ISSUE THREAD

**Issue by linghengqian** [graalvm/native-build-tools#650]
**Describe the bug**
**Make sure that you have read [the documentation](https://graalvm.github.io/native-build-tools) and that you are using the latest plugin version.**

- `[ERROR] Metadata copy process failed with code: 1` is thrown when using `native:metadata-copy` in a multi-module Maven project.

```shell
[INFO] --- native:0.10.4:metadata-copy (default-cli) @ hive-server2-jdbc-driver-uber ---
[DEBUG] Loading mojo org.graalvm.buildtools:native-maven-plugin:0.10.4:metadata-copy from plugin realm ClassRealm[plugin>org.graalvm.buildtools:native-maven-plugin:0.10.4--107866145, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@2626b418]
[DEBUG] Configuring mojo execution 'org.graalvm.buildtools:native-maven-plugin:0.10.4:metadata-copy:default-cli' with basic configurator -->
[DEBUG]   (f) enabled = true
[DEBUG]   (f) defaultMode = Conditional
[DEBUG]   (s) conditional = {userCodeFilterPath=/home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/native-image-filter/user-code-filter.json, parallel=true, extraFilterPath=/home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/native-image-filter/extra-filter.json}
[DEBUG]   (f) modes = org.graalvm.buildtools.maven.config.agent.ModesConfiguration@b548f51
[DEBUG]   (s) disabledStages = [main]
[DEBUG]   (s) merge = false
[DEBUG]   (s) outputDirectory = /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/generated-reachability-metadata/
[DEBUG]   (f) metadataCopy = org.graalvm.buildtools.maven.config.agent.MetadataCopyConfiguration@71f4aeb6
[DEBUG]   (f) agentConfiguration = org.graalvm.buildtools.maven.config.agent.AgentConfiguration@56976b8b
[DEBUG]   (f) project = MavenProject: io.github.linghengqian:hive-server2-jdbc-driver-uber:1.7.0-SNAPSHOT @ /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/hive-server2-jdbc-driver-uber/pom.xml
[DEBUG]   (f) session = org.apache.maven.execution.MavenSession@21bd20ee
[DEBUG] -- end configuration --
[INFO] Copying files from: test
[ERROR] Metadata copy process failed with code: 1

```

**To Reproduce**

- Take Ubuntu WSL 22.04.4 as an example.
- It is assumed that `Git` is configured, and `SDKMAN!` and `Docker Engine` are installed.

```bash
git clone -b add-agent git@github.com:linghengqian/hive-server2-jdbc-driver.git 
cd ./hive-server2-jdbc-driver/
sdk install java 22.0.2-graalce
sdk use java 22.0.2-graalce
./mvnw -PgenerateMetadata -DskipNativeTests clean test native:metadata-copy
```

**Expected behavior**

- There should be no exception.

**Logs**

<details>
<summary>Click me to view the complete Log🦕🐟🐳🐬🐡🦑🦞🐊🐿️🦔🦣</summary>
<pre>

```shell
$ ./mvnw -PgenerateMetadata -DskipNativeTests clean test native:metadata-copy
[INFO] Scanning for projects...
[INFO] Found GraalVM installation from JAVA_HOME variable.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] io.github.linghengqian:hive-parent                                 [pom]
[INFO] hive-server2-jdbc-driver-thin                                      [jar]
[INFO] hive-server2-jdbc-driver-uber                                      [jar]
[INFO] 
[INFO] -----------------< io.github.linghengqian:hive-parent >-----------------
[INFO] Building io.github.linghengqian:hive-parent 1.7.0-SNAPSHOT         [1/3]
[INFO]   from pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ hive-parent ---
[INFO] 
[INFO] --- native:0.10.4:metadata-copy (default-cli) @ hive-parent ---
[WARNING] Cannot find source directory /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/target/native/agent-output/test for metadata copy. Please check if you configured agent properly and it generates all necessary directories. If you want to skipp copy from some source, please configure metadataCopy with disable stage you want to skipp.
[INFO] Metadata copy process finished.
[INFO] 
[INFO] --------< io.github.linghengqian:hive-server2-jdbc-driver-thin >--------
[INFO] Building hive-server2-jdbc-driver-thin 1.7.0-SNAPSHOT              [2/3]
[INFO]   from hive-server2-jdbc-driver-thin/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[WARNING] 2 problems were encountered while building the effective model for org.apache.yetus:audience-annotations:jar:0.5.0 during dependency collection step for project (use -X to see details)
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ hive-server2-jdbc-driver-thin ---
[INFO] Deleting /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/hive-server2-jdbc-driver-thin/target
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ hive-server2-jdbc-driver-thin ---
[INFO] Copying 7 resources from src/main/resources to target/classes
[INFO] 
[INFO] --- compiler:3.13.0:compile (default-compile) @ hive-server2-jdbc-driver-thin ---
[INFO] No sources to compile
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ hive-server2-jdbc-driver-thin ---
[INFO] skip non existing resourceDirectory /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/hive-server2-jdbc-driver-thin/src/test/resources
[INFO] 
[INFO] --- compiler:3.13.0:testCompile (default-testCompile) @ hive-server2-jdbc-driver-thin ---
[INFO] Recompiling the module because of changed source code.
[INFO] Compiling 4 source files with javac [debug release 8] to target/test-classes
[WARNING] 源值 8 已过时，将在未来发行版中删除
[WARNING] 目标值 8 已过时，将在未来发行版中删除
[WARNING] 要隐藏有关已过时选项的警告, 请使用 -Xlint:-options。
[INFO] 
[INFO] --- surefire:3.5.0:test (default-test) @ hive-server2-jdbc-driver-thin ---
[WARNING]  Parameter 'systemProperties' is deprecated: Use systemPropertyVariables instead.
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.thin.ClassExistTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.276 s -- in io.github.linghengqian.hive.server2.jdbc.driver.thin.ClassExistTest
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.thin.StandaloneMetastoreTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 36.40 s -- in io.github.linghengqian.hive.server2.jdbc.driver.thin.StandaloneMetastoreTest
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.thin.ZookeeperServiceDiscoveryTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 66.04 s -- in io.github.linghengqian.hive.server2.jdbc.driver.thin.ZookeeperServiceDiscoveryTest
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.thin.ThinTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 21.95 s -- in io.github.linghengqian.hive.server2.jdbc.driver.thin.ThinTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- native:0.10.4:merge-agent-files (test-native) @ hive-server2-jdbc-driver-thin ---
[INFO] Merging agent 1 files into /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/hive-server2-jdbc-driver-thin/target/native/agent-output/test
[INFO] 
[INFO] --- native:0.10.4:test (test-native) @ hive-server2-jdbc-driver-thin ---
[INFO] Skipping native-image tests (parameter 'skipTests' or 'skipNativeTests' is true).
[INFO] 
[INFO] --- native:0.10.4:metadata-copy (default-cli) @ hive-server2-jdbc-driver-thin ---
[INFO] Copying files from: test
[INFO] Found GraalVM installation from JAVA_HOME variable.
[INFO] Metadata copy process finished.
[INFO] 
[INFO] --------< io.github.linghengqian:hive-server2-jdbc-driver-uber >--------
[INFO] Building hive-server2-jdbc-driver-uber 1.7.0-SNAPSHOT              [3/3]
[INFO]   from hive-server2-jdbc-driver-uber/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ hive-server2-jdbc-driver-uber ---
[INFO] Deleting /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/hive-server2-jdbc-driver-uber/target
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ hive-server2-jdbc-driver-uber ---
[INFO] Copying 7 resources from src/main/resources to target/classes
[INFO] 
[INFO] --- compiler:3.13.0:compile (default-compile) @ hive-server2-jdbc-driver-uber ---
[INFO] No sources to compile
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ hive-server2-jdbc-driver-uber ---
[INFO] skip non existing resourceDirectory /home/linghengqian/TwinklingLiftWorks/git/public/hive-server2-jdbc-driver/hive-server2-jdbc-driver-uber/src/test/resources
[INFO] 
[INFO] --- compiler:3.13.0:testCompile (default-testCompile) @ hive-server2-jdbc-driver-uber ---
[INFO] Recompiling the module because of changed source code.
[INFO] Compiling 4 source files with javac [debug release 8] to target/test-classes
[WARNING] 源值 8 已过时，将在未来发行版中删除
[WARNING] 目标值 8 已过时，将在未来发行版中删除
[WARNING] 要隐藏有关已过时选项的警告, 请使用 -Xlint:-options。
[INFO] 
[INFO] --- surefire:3.5.0:test (default-test) @ hive-server2-jdbc-driver-uber ---
[WARNING]  Parameter 'systemProperties' is deprecated: Use systemPropertyVariables instead.
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.uber.ClassExistTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.224 s -- in io.github.linghengqian.hive.server2.jdbc.driver.uber.ClassExistTest
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.uber.StandaloneMetastoreTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 40.20 s -- in io.github.linghengqian.hive.server2.jdbc.driver.uber.StandaloneMetastoreTest
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.uber.UberTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 27.25 s -- in io.github.linghengqian.hive.server2.jdbc.driver.uber.UberTest
[INFO] Running io.github.linghengqian.hive.server2.jdbc.driver.uber.ZookeeperServiceDiscoveryTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 72.16 s -- in io.github.linghengqian.hive.server2.jdbc.driver.uber.ZookeeperServiceDiscoveryTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- native:0.10.4:merge-agent-files (test-native) @ hive-server2-jdbc-driver-uber ---
[INFO] 
[INFO] --- native:0.10.4:test (test-native) @ hive-server2-jdbc-driver-uber ---
[INFO] Skipping native-image tests (parameter 'skipTests' or 'skipNativeTests' is true).
[INFO] 
[INFO] --- native:0.10.4:metadata-copy (default-cli) @ hive-server2-jdbc-driver-uber ---
[INFO] Copying files from: test
[ERROR] Metadata copy process failed with code: 1
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for io.github.linghengqian:hive-parent 1.7.0-SNAPSHOT:
[INFO] 
[INFO] io.github.linghengqian:hive-parent ................. SUCCESS [  0.121 s]
[INFO] hive-server2-jdbc-driver-thin ...................... SUCCESS [02:08 min]
[INFO] hive-server2-jdbc-driver-uber ...................... FAILURE [02:21 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:31 min
[INFO] Finished at: 2024-12-06T15:46:31+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.graalvm.buildtools:native-maven-plugin:0.10.4:metadata-copy (default-cli) on project hive-server2-jdbc-driver-uber: Metadata copy process failed. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :hive-server2-jdbc-driver-uber
```

</pre>
</details>

**System Info (please complete the following information):**
 - OS: [e.g. `Windows`] `Ubuntu 22.04.4`
 - GraalVM Version [e.g. `22.0 CE`] `GraalVM CE For JDK 22.0.2`
 - Java Version [e.g. `17`] `22.0.2`
 - Plugin version [e.g. `native-gradle-plugin:0.9.10`] `native-maven-plugin:0.10.4`

**Additional context**

- Earlier investigations are at https://github.com/linghengqian/hive-server2-jdbc-driver/pull/16 and https://github.com/apache/shardingsphere/issues/29052 .

**Comment by jvukicev** [comment]
# Implementation proposal

## Scope

Fix Maven `native:metadata-copy` so sequential modules in one reactor can target the same configured metadata directory without a later invocation failing because an earlier module already populated it. Preserve the existing configuration contract: with `merge=false`, each successful invocation replaces the destination metadata, so the final directory deterministically reflects the last applicable module in reactor order; with `merge=true`, the existing destination participates in the next merge. Disabled main/test stages and standard, conditional, direct, and command-line agent enablement retain their current meanings.

The change is limited to the Maven metadata-copy goal, its specification, and repository-owned regression coverage. It will not change agent collection, `native:merge-agent-files`, Gradle metadata-copy behavior, the public Maven configuration model, or the reporter's external project.

## Repository and specification constraints

Update the most-specific Maven behavior specification before production code. Refine `§maven/FS-tracing-agent.4` to define repeated metadata-copy invocations against one destination, including deterministic replacement for `merge=false` and participation of existing metadata for `merge=true`. Keep `§root/FS-tracing-agent.4` and `§root/REQ-backwards-compatibility.2` intact: replacement and merge remain distinct modes, and existing Maven parameters keep their documented meaning.

Add the most-specific marked citations to changed production and functional-test code. If a dedicated issue reproducer is added, update `§maven/E2E-functional-tests.3.6` or `§maven/E2E-functional-tests.3.7` to identify the scenario, following `§maven/AR-maven-plugin.6` and `§root/REQ-real-fixtures`. Keep the implementation Maven-owned under `native-maven-plugin`; do not move build-tool-specific reactor or mojo behavior into `common`, and preserve the existing Native Image executable lookup.

Implementation work must occur in the isolated issue worktree, preserve unrelated changes, and avoid executing the issue-supplied clone, SDK installation, or external build commands. The repository-owned test must provide all project and metadata inputs locally.

## Intended behavior and file changes

In `native-maven-plugin/docs/functional/tracing-agent.md`, extend the merge-and-copy section to define repeated metadata-copy invocations against a shared destination: `merge=false` ignores old metadata and replaces it with the current module's selected stages; `merge=true` reads the old destination as an additional input and replaces it with the combined result.

In `native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/MetadataCopyMojo.java`, make repeated sequential reactor invocations safe without changing `MetadataCopyConfiguration` or treating `merge=false` as an implicit merge. If `native-image-configure` cannot write to a populated destination, direct it to a temporary output directory and replace the configured destination only after it succeeds. For merge mode, continue to validate and include the existing destination as an input; for replace mode, use only the current module's selected stage directories.

Keep the implementation private to the Maven plugin and preserve current diagnostics and Native Image executable lookup. Only add stronger rollback, interruption, or filesystem-transaction behavior if implementation demonstrates that it is necessary for a safe replacement on supported platforms.

Add `native-maven-plugin/reproducers/issue-650/` as a minimal parent with at least two leaf modules sharing one explicit output directory. Give each module valid, distinguishable test-stage reachability metadata and configure main-stage copying off. Avoid network services, Docker, SDK managers, and the issue-linked repository.

Add a focused Maven functional test under `native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/issues/`, using the issue reproducer. Verify that a reactor invocation reaches both leaf modules and succeeds with `merge=false`, that the final shared destination contains the later module's marker and not the earlier module's marker, and that rerunning the goal remains deterministic. Add a merge-mode case using the same fixture or a focused variant to prove both modules' distinguishable metadata survive when `merge=true`.

Update `native-maven-plugin/docs/e2e.md` only as needed to register the new issue scenario in the appropriate functional-test section. No user documentation or changelog change is required unless implementation discovers a user-visible configuration or diagnostic change beyond the clarified existing contract.

## Validation strategy

First validate the focused specification and citation changes with workspace-wide grund checking. Exercise the new issue-650 functional test on a supported GraalVM so it invokes the packaged Maven plugin and real `native-image-configure` across the multi-module reactor. Inspect the resulting shared directory after replace mode, a repeated replace run, and merge mode to verify exact marker ownership and deterministic contents.

Run the existing tracing-agent and metadata-copy functional scenarios, including disabled-stage and conditional-mode coverage, to ensure the change does not alter source selection, enablement, or default destinations. Run focused unit coverage only if the implementation introduces a testable helper, then the complete Maven functional suite and Maven plugin inspections. Because production behavior remains Maven-specific and no shared or Gradle code should change, Gradle functional validation is unnecessary unless implementation unexpectedly crosses that boundary.

If staging is used, validate that a failed post-processing invocation reports a Maven goal failure and does not leave staging artifacts. Review the target worktree diff to ensure only the scoped Maven spec, implementation, tests, and local reproducer changed.

## Risks and mitigations

- Replacing a user-selected directory can lose unrelated files if the directory is not exclusively metadata. Treat the configured `outputDirectory` according to its existing replacement contract and document the behavior precisely.
- Merge mode currently uses the destination as both input and output. Separating those paths could expose assumptions in `native-image-configure`; functional coverage must verify the supported metadata formats through the existing validation rules.
- Parallel Maven reactors can create simultaneous writers to one explicit destination. Use unique staging paths so temporary files do not collide, but do not claim deterministic concurrent-writer ordering without synchronization and dedicated coverage; the reported sequential reactor case must remain the primary acceptance target.
- Cross-platform replacement and cleanup behavior differs, especially on Windows. Prefer Java NIO operations and rely on the Maven functional matrix across Ubuntu, Windows, and macOS.
- A test that merely observes a successful exit could miss accidental merging in replace mode. Use distinguishable valid metadata and assert both presence and absence at the final destination for each mode.

## Known gaps

The issue's external reproducer and linked draft pull request have not been executed or used as implementation instructions. Current-master failure still needs to be confirmed with the repository-owned multi-module fixture, and the exact `native-image-configure` diagnostic behind exit code 1 may vary across supported GraalVM versions.

The accepted scope defines deterministic behavior for sequential reactor modules. It does not yet define a stable winner for genuinely concurrent modules writing one destination; if the repository-owned test demonstrates that supported parallel Maven execution must also be serialized, destination locking would require an explicit specification extension and additional review.

The precise internal helper boundaries and any temporary-output naming are implementation details. They should remain private to the Maven plugin unless testing demonstrates that a reusable abstraction is warranted.

---

Proposal ID: `226e39d746e331fd`

Approve:

```text
/rhei approve 226e39d746e331fd
```

Reject with an explanation on following lines:

```text
/rhei reject 226e39d746e331fd
<explain what should change>
```

This implementation proposal was generated by AI using `openai:gpt-5.6-sol` through [Rhei](https://github.com/vjovanov/rhei).
